tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/core/NN.java

```diff
@@ -0,0 +1,379 @@
+   *
+   * <p>Measures the probability error in discrete classification tasks in which the classes are
+   * mutually exclusive (each entry is in exactly one class). For example, each CIFAR-10 image is
+   * labeled with one and only one label: an image can be a dog or a truck, but not both.
+   *
+   * <p>**NOTE:** While the classes are mutually exclusive, their probabilities need not be. All
+   * that is required is that each row of `labels` is a valid probability distribution. If they are
+   * not, the computation of the gradient will be incorrect.
+   *
+   * <p>If using exclusive `labels` (wherein one and only one class is true at a time), see
+   * `sparse_softmax_cross_entropy_with_logits`.
+   *
+   * <p>Usage:
+   *
+   * <pre>
+   *   >>> logits = [[4.0, 2.0, 1.0], [0.0, 5.0, 1.0]]
+   *   >>> labels = [[1.0, 0.0, 0.0], [0.0, 0.8, 0.2]]
+   *   >>> tf.nn.softmax_cross_entropy_with_logits(labels=labels, logits=logits)
+   *   <tf.Tensor: shape=(2,), dtype=float32,
+   *   numpy=array([0.16984604, 0.82474494], dtype=float32)>
+   * </pre>
+   *
+   * <p>Backpropagation will happen into both `logits` and `labels`. To disallow backpropagation
+   * into `labels`, pass label tensors through `tf.stop_gradient` before feeding it to this
+   * function.
+   *
+   * @param scope current scope
+   * @param labels Each vector along the class dimension should hold a valid probability
+   *     distribution e.g. for the case in which labels are of shape `[batch_size, num_classes]`,
+   *     each row of `labels[i]` must be a valid probability distribution.
+   * @param logits Per-label activations, typically a linear output. These activation energies are
+   *     interpreted as unnormalized log probabilities.
+   * @param axis The class dimension. -1 is the last dimension.
+   * @param <U> the data type of the logits
+   * @param <T> the number type of the operands
+   * @return the softmax cross entropy loss. Its type is the same as `logits` and its shape is the
+   *     same as `labels` except that it does not have the last dimension of `labels`.
+   */
+  @Endpoint(name = "softmaxCrossEntropyWithLogits")
+  public static <U extends TType, T extends TNumber> Operand<T> softmaxCrossEntropyWithLogits(
+      Scope scope, Operand<T> labels, Operand<U> logits, int axis) {
+    axis = axis % logits.asOutput().shape().numDimensions();
+    if (axis < 0) {
+      axis += logits.asOutput().shape().numDimensions();
+    }
+
+    Operand precise_logits =
+        logits; // cannot use generics cause logits of bool gets cast to TFloat32
+
+    boolean convertToFloat32 =
+        logits.asOutput().dataType() == TFloat16.DTYPE
+            || logits.asOutput().dataType() == TBfloat16.DTYPE;
+    if (convertToFloat32) {
+      precise_logits = Cast.create(scope, logits, TFloat32.DTYPE);
+    }
+    /* cannot use generics on DataType because precis_logits may have been cast. */
+    DataType dtype = precise_logits.asOutput().dataType();
```

Comment:
I have tried all kinds of combinations, `<?>`, `<? extends TType> or <? extends TNumber>`, etc.,  but the main hangup is when I call other methods that have generics defined as in `<T extends TType>`, it won't compile. The only way I can get it to compile is to remove the generic.
Do you have any other suggestions?

Reply:
Interesting. You are right that `<?>` is a bad pick, we need at least to be bound to `TType`, so `<? extends TType>`. Now you are saying that even with this generic, you are not able to call a method that accepts only `<T extends TType>`? Do you have an example?

Reply:
Here is a code fragment that highlights the issue. 

```
public static <T extends TNumber> Operand<T> foo(Ops tf, Operand<T> logits) {
        Operand<? extends TNumber> preciseLogits = tf.dtypes.cast(logits, TFloat32.DTYPE);
        return flattenOuterDims(tf.scope(), preciseLogits);
}
    
private static <T extends TNumber> Operand<T> flattenOuterDims(Scope scope, Operand<T> logits) {
    return null;
}
```
I get compile error on the line `return flattenOuterDims`:

   ```
GenericExample.java:[35,32] incompatible types: inferred type does not conform to equality constraint(s)
    inferred: T
    equality constraints(s): T,capture#1 of ? extends org.tensorflow.types.family.TNumber
```

BTW: I have also run into issues with consistency of `<T>` and `<U>`. Usually, `<T extends TType>` and `<U extends TNumber>`, but some times I run into methods where `<T extends TNumber>`. This causes conflicts similar to what I see with the above code. Seems that the compiler is looking for consistency in the generic definitions.

If I change the above  to:
`Operand<T> preciseLogits = tf.dtypes.cast(logits, TFloat32.DTYPE);`

I get an error complaining mismatched generics on <T> and <U> on the line that does the line doing the cast.

```
GenericExample.java:[34,50] incompatible types: inferred type does not conform to equality constraint(s)
    inferred: T
    equality constraints(s): T,org.tensorflow.types.TFloat32
```

Reply:
Ok for this example specifically, I can see that it won't work because the type of the returned value in `foo` is `T`, while the one returned by `flattenOuterDims` is `<? extends Number>` (inferred by `preciseLogits`). Both `T` are unrelated.

It seems that in your code, you are always expecting to return a `TFloat32` tensor as the result of `foo` with the explicit cast. Therefore, you should enforce this in the signature of the method as well, i.e.
```java
public static <T extends TNumber> Operand<TFloat32> foo(Ops tf, Operand<T> logits) {
        Operand<TFloat32> preciseLogits = tf.dtypes.cast(logits, TFloat32.DTYPE);
        return flattenOuterDims(tf.scope(), preciseLogits);
}
    
private static <T extends TNumber> Operand<T> flattenOuterDims(Scope scope, Operand<T> logits) {
    return null;
}
```

Reply:
This was a quick example, I could make this change with these methods which are under my control, but I will still run into the underlying issue when I call other apis in `tensorflow.core-api`. The real code in `org.tensorflow.op.nn.SoftmaxCrossEntropyWithLogits` only does the cast to `TFloat32` when it is 16  bit Float type (`TBFLoat16` or `TFloat16`). No cast is required, If is is already `TFLoat32`, `TFloat64`, `TInt32`, or `TInt64`.

In `org.tensorflow.op.nn.SoftmaxCrossEntropyWithLogits`, if I change to `Operand<? extends TNumber> preciseLogits`, then I get an error when calling:
```
org.tensorflow.op.nn.raw.SoftmaxCrossEntropyWithLogits smax =
        org.tensorflow.op.nn.raw.SoftmaxCrossEntropyWithLogits.create(
            scope, preciseLogits, castLabels);
```
`'create(org.tensorflow.op.Scope, org.tensorflow.Operand<T>, org.tensorflow.Operand<T>)' in 'org.tensorflow.op.nn.raw.SoftmaxCrossEntropyWithLogits' cannot be applied to '(org.tensorflow.op.Scope, org.tensorflow.Operand<capture<? extends org.tensorflow.types.family.TNumber>>, org.tensorflow.Operand<capture<? extends org.tensorflow.types.family.TNumber>>)'`

Reply:
In this case, the problem is that the signature of `raw.SoftmaxCrossEntropyWithLogits` enforces that both `preciseLogits` and `castLabels` should be of the same type. When I look at your [current implementation](https://github.com/JimClarke5/java/blob/b8d3ac2d001251c95bd55ed9cd902431108468bd/tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/nn/SoftmaxCrossEntropyWithLogits.java#L82), this information gets lost so the compiler fails (and TF runtime will also fail if this condition is not met). 

I see a few things that could help, if you want to give it a try:
* `U` must be bound to `TNumber` as well, since you assert it later with an explicit cast
* You can get rid of the casting complexity for float types with a recursively call the same method when required, so the method could recapture the effective type:
```java
    if (logits.asOutput().dataType() == TFloat16.DTYPE || logits.asOutput().dataType() == TBfloat16.DTYPE) {
        return softmaxCrossEntropyWithLogits(scope, labels, Cast.create(scope, logits, TFloat32.DTYPE));
    }
```
* Make sure all parameterized variables keeps track of their type (`DataType`, `Operand`, ...).

Reply:
Based on your suggestion, I did the following, using recursion, which does compile. I still need to validate that I don't get a runtime error.

```
public static <T extends TNumber> Operand<T> softmaxCrossEntropyWithLogits(
      Scope scope, Operand<T> labels, Operand<T> logits, int axis) {

...
if (convertToFloat32) {
      Operand<TFloat32> result =  softmaxCrossEntropyWithLogits(scope,
              Cast.create(scope, labels, TFloat32.DTYPE),
              Cast.create(scope, logits, TFloat32.DTYPE),
              axis);
      return Cast.create(scope, result, logits.asOutput().dataType());
} else if(!logits.asOutput().dataType().equals(labels.asOutput().dataType())) {
      return softmaxCrossEntropyWithLogits(scope,
              Cast.create(scope, labels, logits.asOutput().dataType()),
              logits,
              axis);
}
....
```

Reply:
I did have to cast labels when calling  `org.tensorflow.op.nn.raw.SoftmaxCrossEntropyWithLogits.create`

At this point, both args will be the same datatype.

```
org.tensorflow.op.nn.raw.SoftmaxCrossEntropyWithLogits<T> smax =
        org.tensorflow.op.nn.raw.SoftmaxCrossEntropyWithLogits.create(
            scope, logits, (Operand<T>)labels);
```

Reply:
I had to change the signature to:

```
public static <T extends TNumber, U extends TNumber> Operand<T> softmaxCrossEntropyWithLogits(
      Scope scope, Operand<U> labels, Operand<T> logits, int axis) {
```

This is because `labels` and `logits` may be different types initially.